chainstate/src/detail/transaction_verifier/mod.rs

```diff
@@ -489,6 +577,212 @@ impl<'a, S: BlockchainStorageRead> TransactionVerifier<'a, S> {
+        match self.tokens_cache.entry(token_id) {
+            Entry::Occupied(entry) => {
+                let tokens_op = entry.into_mut();
+                *tokens_op = CachedTokensOperation::Erase;
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(CachedTokensOperation::Erase);
+            }
+        }
+        Ok(())
+    }
+
+    fn check_connected_tokens(
+        &mut self,
+        token_data: &TokenData,
+        tx: &Transaction,
+        block: &Block,
+    ) -> Result<(), TokensError> {
+        match token_data {
+            TokenData::TokenTransferV1 { token_id, amount } => {
+                self.check_connected_transfer_data(block.get_id(), tx, token_id, amount)?;
+            }
+            TokenData::TokenIssuanceV1 {
+                token_ticker,
+                amount_to_issue,
+                number_of_decimals,
+                metadata_uri,
+            } => {
+                self.check_connected_issue_data(
+                    token_ticker,
+                    amount_to_issue,
+                    number_of_decimals,
+                    metadata_uri,
+                    tx.get_id(),
+                    block.get_id(),
+                )?;
+                self.register_issuance(tx)?;
+            }
+            TokenData::TokenBurnV1 {
+                token_id,
+                amount_to_burn,
+            } => {
+                self.check_connected_burn_data(tx, &block.get_id(), token_id, amount_to_burn)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn check_tokens_values(&mut self, block: &Block) -> Result<(), ConnectTransactionError> {
+        for tx in block.transactions() {
+            // Check tokens before connect tx
+            tx.outputs()
+                .iter()
+                .filter_map(|output| match output.value() {
+                    OutputValue::Coin(_) => None,
+                    OutputValue::Token(token_data) => Some(token_data),
+                })
+                .try_for_each(|token_data| self.check_connected_tokens(token_data, tx, block))
+                .map_err(ConnectTransactionError::TokensError)?;
+
+            if self.is_issuance_in_custom_block(tx)? {
+                // check is fee enough for issuance
+                ensure!(
+                    self.check_transferred_amounts_and_get_fee(tx).map_err(|_| {
+                        ConnectTransactionError::TokensError(TokensError::InsufficientTokenFees(
+                            tx.get_id(),
+                            block.get_id(),
+                        ))
+                    })? >= TOKEN_MIN_ISSUANCE_FEE,
+                    ConnectTransactionError::TokensError(TokensError::InsufficientTokenFees(
+                        tx.get_id(),
+                        block.get_id()
+                    ),)
+                );
+            }
+        }
+        Ok(())
+    }
+
+    fn is_issuance_in_custom_block(
```

Comment:
What does "custom" mean? Please refrain from introducing arbitrary terms. If what you mean here is "non_genesis", then call it that. 

In fact, I'm really having trouble understanding the purpose of this. Why does it matter if the current best is genesis or not?

Reply:
Depends on, will we have any tokens in genesis or not?

Reply:
If we don't plan to have tokens in genesis, then we can simplify this check

Reply:
We can't have tokens on genesis because there are no inputs, meaning, there's no systematic way to calculate any token id.

Reply:
Fixed, removed unnecessary check.